zero_functional.nim

```diff
@@ -843,12 +855,111 @@ proc inlineElement(ext: ExtNimNode): ExtNimNode {.compileTime.} =
+  while idx != items.currentIdx and items.currentIt != nil:
+    items.currentIt = items.currentIt.next
+    items.currentIdx += 1
+  if items.currentIt == nil:
+    raise newException(StopIteration, "index too large!")
+  items.currentIdx += 1
+  let value = items.currentIt.value
+  items.currentIt = items.currentIt.next
+  result = value
+
+proc high*(items: MkListIndexable): int =
+  # high = is broken down to number of items - 1
+  # so that low would be mapped to 0
+  if items.length == -1:
+    for it in items.listRef:
+      items.length += 1 
+  return items.length
+
+proc `[]`*[T] (items: MkSliceIndexable[T], idx: Natural): T =
+  items.slice.a + T(idx)
+
+proc high*(items: MkSliceIndexable): int =
+  int(items.slice.b - items.slice.a)
+
+## Wraps the given node with `mkIndexable` when the node type does not support access with `[]`
+proc wrapIndexable(a: NimNode): NimNode {.compileTime.} =
+  let q = quote:
+    when not compiles(`a`[0]) or not compiles(high(`a`)):
+      when not compiles(mkIndexable(`a`)):
+        static:
+          assert(false, "need to provide an own implementation for mkIndexable(" & $`a`.type & ")")
+      else:
+        var `a` = mkIndexable(`a`)
+  result = q
+
+## Replaces zip(a,b,c) --> ... with something like
+## a --> filter(idx <= min([b.high(),c.high()]) --> map(it,b[idx],c[idx])
+## Types that do not support `high` and `[]` have to implement the function
+## `mkIndexable`(`MyType`) returning a type that supports both functions.
+## This should be done, when the type cannot simply be mapped to `[]` and `high` -
+## see `MkListIndexable`.
+proc replaceZip(args: NimNode) : NimNode {.compileTime.} =
+  var idx = 0
+  result = nnkStmtList.newTree()
+
+  for arg in args:
+    # search for all zip calls and replace them with filter --> map
+    if arg.kind == nnkCall and arg[0].label == $Command.zip:
+      let zipCmd = arg.copyNimTree()
+      let highList = nnkBracket.newTree()
+      let params = newPar()
+      if idx == 0:
+        params.add(newIdentNode(iteratorVariableName)) # map(it)
+      for paramIdx in params.len+1..<zipCmd.len:
+        let p = zipCmd[paramIdx]
+        if p.kind != nnkInfix and p.kind != nnkBracketExpr and p.label != iteratorVariableName:
+          result.add(p.wrapIndexable())
+          highList.add(nnkCall.newTree(newIdentNode("high"), p))
+          params.add(nnkBracketExpr.newTree(p, newIdentNode(indexVariableName))) # map(it,b[idx],...)
+        else:
+          params.add(p)
+      if idx == 0:
+        # set the first arg of zip as the first arg in the chain (e.g. moving `a` to the left of `-->`)
+        args.insert(0, zipCmd[1]) # a --> ...
+        idx = 1
+      if highList.len > 0:
+        let minHigh = genSym(nskLet, "__minHigh__")
+        let i = quote:
+          let `minHigh`= min(`highList`)
+        result.add(i)
+        let filterCmd = newCall($Command.filter, infix(newIdentNode(indexVariableName), "<=", minHigh))  
```

Comment:
it does not seem to be a performance penalty using filter here - at least not on my laptop. On the other hand this approach of converting zip(a,b,c) to a --> filter(...) --> map(...) is a more general approach than say maybe 
```nim
for idx in maxLow...minHigh:
  let it0 = (a[idx], b[idx], c[idx])
``` 
although it might not look like this - but
+ `a` does not need to implement the `[]` and `high`
+ it is possible to use `zip` in-between (still looking for use-cases) 
+ no special rules required when generating the overall result
... hm - but maybe there is a way of keeping it general with the ease of the other for loop...